File: holo-e2e-performance-tool/src/main/java/com/alibaba/hologres/performace/client/ScanTest.java
package com.alibaba.hologres.performace.client;


import com.alibaba.hologres.client.*;
import com.alibaba.hologres.client.impl.ConnectionHolder;
import com.alibaba.hologres.client.impl.ExecutionPool;
import com.alibaba.hologres.client.impl.util.ConnectionUtil;
import com.alibaba.hologres.client.model.RecordScanner;
import com.alibaba.hologres.client.model.TableName;
import com.alibaba.hologres.client.model.TableSchema;
import com.alibaba.hologres.client.utils.ConfLoader;
import com.alibaba.hologres.client.utils.Metrics;
import com.alibaba.hologres.com.codahale.metrics.Histogram;
import com.alibaba.hologres.com.codahale.metrics.Meter;
import com.alibaba.hologres.org.postgresql.PGProperty;
import com.alibaba.hologres.performace.params.ParamsProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.DriverManager;
import java.util.Properties;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class ScanTest {
  public static final Logger LOG = LoggerFactory.getLogger(ScanTest.class);
  public static final String METRICS_SCAN_PERF_QPS = "scan_perf_qps";
  public static final String METRICS_SCAN_PERF_LATENCY = "scan_perf_latency";

  private String confName;
  private long targetTime;
  private AtomicInteger singleExecutionPoolJobSize;
  ScanTestConf conf = new ScanTestConf();

  ParamsProvider provider;

  private static long memoryUsage = 0;
  private CyclicBarrier barrier = null;

  public void run(String confName) throws Exception {
    LOG.info("confName:{}", confName);
    this.confName = confName;
    ConfLoader.load(confName, "scan.", conf);

    HoloConfig config = new HoloConfig();
    ConfLoader.load(confName, "holoClient.", config);
    provider = new ParamsProvider(conf.keyRangeParams);
    Reporter reporter = new Reporter(confName);
    ConnectionHolder.addPreSql("set hg_experimental_enable_fixed_dispatcher_for_scan = on");
    Properties props = new Properties();
    PGProperty.USER.set(props, config.getUsername());
    PGProperty.PASSWORD.set(props, config.getPassword());
    PGProperty.APPLICATION_NAME.set(props, config.getAppName());
    LOG.info("props : {}", props);
    String jdbcUrl = config.getJdbcUrl();
    try (Connection conn = DriverManager.getConnection(jdbcUrl, props)) {
      if (conf.vacuumTableBeforeRun) {
        SqlUtil.vaccumTable(conn, conf.tableName);
      }
      reporter.start(ConnectionUtil.getHoloVersion(conn));
      TableSchema schema = ConnectionUtil.getTableSchema(conn, TableName.valueOf(conf.tableName));
      if (schema == null) {
        throw new Exception("table not found");
      } else if (schema.getDistributionKeys().length != provider.size()) {
        throw new Exception(
            "table has " + schema.getDistributionKeys().length + " distribution keys, but test.params only has "
                + provider.size() + " columns");
      }
    }
    if (conf.singleExecutionPool) {
      singleExecutionPoolJobSize = new AtomicInteger(conf.threadSize);
    } else {
      singleExecutionPoolJobSize = new AtomicInteger(0);
    }
    barrier = new CyclicBarrier(conf.threadSize, ()->{
      memoryUsage = Util.getMemoryStat();
      Util.dumpHeap(confName);
    });
    targetTime = System.currentTimeMillis() + conf.testTime;
    Thread[] threads = new Thread[conf.threadSize];
    Metrics.startSlf4jReporter(60L, TimeUnit.SECONDS);
    for (int i = 0; i < threads.length; ++i) {
      threads[i] = new Thread(new Job(i));
      threads[i].start();
    }

    for (int i = 0; i < threads.length; ++i) {
      threads[i].join();
    }

    Metrics.reporter().report();
    {
      Meter meter = Metrics.registry().meter(METRICS_SCAN_PERF_QPS);
      Histogram hist = Metrics.registry().histogram(METRICS_SCAN_PERF_LATENCY);
      reporter.report(meter.getCount(), meter.getOneMinuteRate(), meter.getFiveMinuteRate(),
          meter.getFifteenMinuteRate(), hist.getSnapshot().getMean(),
          hist.getSnapshot().get99thPercentile(), hist.getSnapshot().get999thPercentile(), memoryUsage);
    }

    if (conf.deleteTableAfterDone) {
      SqlUtil.dropTableByHoloClient(config, conf.tableName);
    }
  }


  class Job implements Runnable {
    int id;

    public Job(int id) {
      this.id = id;
    }

    @Override
    public void run() {
      HoloConfig poolConf = new HoloConfig();
      HoloConfig clientConf = new HoloConfig();
      HoloClient client = null;
      HoloClientExecutionPool pool = null;
      try {
        ConfLoader.load(confName, "holoClient.", poolConf);
        ConfLoader.load(confName, "holoClient.", clientConf);
        ConfLoader.load(confName, "pool.", poolConf);

        Meter meter = Metrics.registry().meter(METRICS_SCAN_PERF_QPS);
        Histogram hist = Metrics.registry().histogram(METRICS_SCAN_PERF_LATENCY);
        client = new HoloClient(clientConf);
        pool = new HoloClientExecutionPool(poolConf, this.id, conf.singleExecutionPool);
        pool.setHoloClientPool(client);
        int i = 0;
        CompletableFuture<Void> future = null;
        while (true) {
          if (++i % 1000 == 0) {
            if (System.currentTimeMillis() > targetTime) {
              break;
            }
          }
          TableSchema schema = client.getTableSchema(conf.tableName);
          Scan.Builder scanBuilder = Scan.newBuilder(schema).setSortKeys(SortKeys.NONE);
          for (int j = 0; j < schema.getDistributionKeys().length; ++j) {
            scanBuilder.addEqualFilter(schema.getDistributionKeys()[j], provider.get(j));
          }
          Scan scan = scanBuilder.build();

          long startNano = System.nanoTime();
          if (conf.async) {
            future = client.asyncScan(scan).thenAccept(rs -> {
              meter.mark();
              long endNano = System.nanoTime();
              hist.update((endNano - startNano) / 1000000L);
            });
          } else {
            try (RecordScanner rs = client.scan(scan)) {
            }
            meter.mark();
            long endNano = System.nanoTime();
            hist.update((endNano - startNano) / 1000000L);
          }
        }
        if (conf.async && future != null) {
          future.get();
        }
      } catch (Exception e) {
        e.printStackTrace();
      } finally {
        if (conf.dumpMemoryStat) {
          try {
            barrier.await();
          } catch (Exception e) {
            e.printStackTrace();
          }
        }
        if (client != null) {
          client.close();
        }
        if (conf.singleExecutionPool && singleExecutionPoolJobSize.decrementAndGet() == 0 || (!conf.singleExecutionPool)) {
          if (pool != null) {
            pool.close();
          }
        }
      }
    }
  }
}

class ScanTestConf {
  public int threadSize = 10;
  public long testTime = 600000;
  public String tableName = "holo_perf";
  public boolean singleExecutionPool = true;
  public boolean vacuumTableBeforeRun = true;
  public String keyRangeParams;
  public boolean async;
  public boolean deleteTableAfterDone = false;
  public boolean dumpMemoryStat = false;
}

File: holo-client-py/perf/scan_test.py
"""SCAN mode prefix scan test.

Executes SELECT * FROM table WHERE id = ? using HoloClient.scan() API.
Each thread uses its own HoloClient instance.
Mirrors Java ScanTest.
"""

from __future__ import annotations

import logging
import threading
import time
from typing import List

import psycopg

from hologres import HoloClient, Scan, SortKeys

from .conf_loader import build_holo_config, load_conf
from .config import ScanTestConf
from .metrics import Histogram, Meter
from .params import ParamsProvider
from .reporter import Reporter
from .sql_util import vacuum_table

logger = logging.getLogger(__name__)


class ScanTest:
    """Prefix scan benchmark using HoloClient.scan() API."""

    def __init__(self):
        self.conf = ScanTestConf()
        self.holo_config = None
        self.target_time = 0.0
        self.meter = Meter()
        self.histogram = Histogram()

    def run(self, conf_name: str) -> None:
        load_conf(conf_name, "scan.", self.conf)
        self.holo_config = build_holo_config(conf_name)

        provider = ParamsProvider(self.conf.key_range_params)

        # Setup
        if self.conf.vacuum_table_before_run:
            with psycopg.connect(self.holo_config.conninfo, autocommit=True) as conn:
                vacuum_table(conn, self.conf.table_name)

        logger.info(
            "SCAN test: %d threads, table=%s",
            self.conf.thread_size,
            self.conf.table_name,
        )

        reporter = Reporter(conf_name)
        reporter.start(self.holo_config.conninfo)
        self.target_time = time.time() + self.conf.test_time / 1000.0

        # Spawn threads
        threads: List[threading.Thread] = []
        for i in range(self.conf.thread_size):
            t = threading.Thread(
                target=self._scan_job,
                args=(i, provider),
                name=f"Thread-{i}",
                daemon=True,
            )
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        logger.info("finished, %d scans executed", self.meter.count)
        reporter.report(
            self.meter, self.histogram, dump_memory_stat=self.conf.dump_memory_stat
        )

        # Cleanup
        if self.conf.delete_table_after_done:
            with psycopg.connect(self.holo_config.conninfo, autocommit=True) as conn:
                from .sql_util import drop_table

                drop_table(conn, self.conf.table_name)

    def _scan_job(self, thread_id: int, provider: ParamsProvider) -> None:
        """Each thread creates its own HoloClient and runs scan queries."""
        try:
            client = HoloClient(self.holo_config)
            schema = client.get_table_schema(self.conf.table_name)
            meter = self.meter
            histogram = self.histogram
            i = 0

            while True:
                i += 1
                if i % 100 == 0 and time.time() > self.target_time:
                    break

                key_value = provider.get(0)

                scan = (
                    Scan.builder(schema)
                    .add_equal_filter("id", key_value)
                    .set_sort_keys(SortKeys.NONE)
                    .build()
                )

                start_ns = time.monotonic_ns()
                client.scan(scan)
                elapsed_ms = (time.monotonic_ns() - start_ns) / 1_000_000

                meter.mark()
                histogram.update(elapsed_ms)

            logger.info("Thread-%d: %d scans", thread_id, i)
            client.close()
        except Exception:
            logger.exception("Thread-%d failed", thread_id)
